### PR TITLE
Fix problems in octree search functions when using dynamic depth

### LINKED ISSUE #1232: crash in nearestKSearch when using enableDynamicDepth
If I build an octree using enableDynamicDepth, the nearestKSearch crashes with an access violation.  It seems to be mis-identifying the node type, thinking that a leaf node is a branch node, and trying too traverse too deeply.  Avoiding enableDynamicDepth prevents the crash.

In octree_search.hpp,:
pcl::octree::OctreePointCloudSearch<PointT, LeafContainerT, BranchContainerT>::getKNearestNeighborRecursive ...

``` cpp
    if (tree_depth < this->octree_depth_)   // Doesn't the node know if it's is a branch/leaf ? 
    {
      // we have not reached maximum tree depth
      smallest_squared_dist = getKNearestNeighborRecursive (point, K, static_cast<const BranchNode*> (child_node), new_key, tree_depth + 1,
                                                            smallest_squared_dist, point_candidates);
    }
    else
    {
      // we reached leaf node level
```

### LINKED ISSUE #4591: Octree: Segfault when using DynamicDepth + boxSearch
I am attempting to perform a box search on an Octree that has enabled dynamic depth. This is producing a segfault that is resolved if I do not enable dynamic depth. I am using Pcl-1.11.1 built with gcc-8.3.1 on Power9 architecture. 

Minimal example required to reproduce result:
```
#include <vector>
#include <pcl/point_cloud.h>
#include <pcl/octree/octree.h>

int main() {
    constexpr float resolution = 0.01f;
    constexpr std::size_t leafAggSize = 5;
    pcl::octree::OctreePointCloudSearch<pcl::PointXYZ> octree (resolution);
    pcl::PointCloud<pcl::PointXYZ>::Ptr new_cloud (new pcl::PointCloud<pcl::PointXYZ> ());
    octree.setInputCloud (new_cloud);
    std::vector<pcl::PointXYZ> points = {pcl::PointXYZ(1.08512, 8.55778, -22.1631), pcl::PointXYZ(-2.19884, -3.77412, 2.63825), pcl::PointXYZ(8.6194, -17.5698, -49.5281), pcl::PointXYZ(-8.58222, -18.9215, -31.3533), pcl::PointXYZ(4.26873, -14.4946, 32.5853), pcl::PointXYZ(-1.1815, -18.1647, 37.0143), pcl::PointXYZ(1.87733, -21.8159, 39.1322), pcl::PointXYZ(3.1078, -14.5399, 2.33477), pcl::PointXYZ(-7.86679, 3.11481, -39.1623), pcl::PointXYZ(-12.3546, -14.0151, -19.2577), pcl::PointXYZ(11.9656, 22.5092, 31.1683), pcl::PointXYZ(-9.33364, -16.4029, -42.1012), pcl::PointXYZ(7.90562, -9.43234, -22.5926), pcl::PointXYZ(3.30959, -3.41479, 19.9359), pcl::PointXYZ(11.0007, 7.09825, 31.7649)};
    for (int i = 0; i < points.size(); i++)
    {
      new_cloud->push_back (points[i]);
    }
    octree.defineBoundingBox (-100, -100, -100, 100, 100, 100);
    octree.enableDynamicDepth (leafAggSize);
    octree.addPointsFromInputCloud ();
    Eigen::Vector3f bbox_min = Eigen::Vector3f (0,0,0);
    Eigen::Vector3f bbox_max = Eigen::Vector3f (2, 3, 4);

    std::vector<int> result_indices;
    std::cout << "about to bbox search" << std::endl;
    octree.boxSearch (bbox_min, bbox_max, result_indices);
    for(int i = 0; i < result_indices.size(); i++) {
        std::cout << "early find found index: " << result_indices[i] << std::endl;
    }
}
```
The segfault persists even if I set defineBoundingBox() to bounds that fully encompass the query bounding box. Removing the line `octree.enableDynamicDepth (leafAggSize);` causes the query to work properly with no segfault. Other sets of input points work properly, but the 15 used here are representative of the type that I am working with (can be positive or negative in x,y,z).

### PR DESCRIPTION
**Pull request by mvieth**
Please see commit messages.
Fixes #1232  Fixes #4591

### CODE REVIEW

File: test/octree/test_octree.cpp [hunk 1]
```diff
@@ -1142,80 +1142,84 @@ TEST (PCL, Octree_Pointcloud_Nearest_K_Neighbour_Search)
...
-    cloudIn->width = 1000;
-    cloudIn->height = 1;
-    cloudIn->points.resize (cloudIn->width * cloudIn->height);
-    for (std::size_t i = 0; i < 1000; i++)
-    {
-      (*cloudIn)[i] = PointXYZ (static_cast<float> (5.0  * rand () / RAND_MAX),
-                                     static_cast<float> (10.0 * rand () / RAND_MAX),
-                                     static_cast<float> (10.0 * rand () / RAND_MAX));
-    }
+      // generate point cloud
```
Comment (larshg) [inline test/octree/test_octree.cpp:1158]: Seems a lot of places, we could use CloudGenerator in this test file? Should we add it as a good first issue?

File: test/octree/test_octree.cpp [hunk 2]
```diff
@@ -1367,83 +1379,85 @@ TEST (PCL, Octree_Pointcloud_Neighbours_Within_Radius_Search)
...
+    for (unsigned int test_id = 0; test_id < test_runs; test_id++)
     {
-      (*cloudIn)[i] = PointXYZ (static_cast<float> (10.0 * rand () / RAND_MAX),
-                                     static_cast<float> (10.0 * rand () / RAND_MAX),
-                                     static_cast<float> (5.0  * rand () / RAND_MAX));
-    }
+      // define a random search point
+      PointXYZ searchPoint (static_cast<float> (10.0 * rand () / RAND_MAX),
+                            static_cast<float> (10.0 * rand () / RAND_MAX),
+                            static_cast<float> (10.0 * rand () / RAND_MAX));
```
Comment (larshg) [inline test/octree/test_octree.cpp:1391]: Reason why Z value should be further out than the generated test cloud? (10.0 vs 5.0)

**Review by larshg — APPROVED** [review]

File: test/octree/test_octree.cpp (hunk 2 above)
Comment (mvieth) [reply test/octree/test_octree.cpp:1391]: Yes, probably.

File: test/octree/test_octree.cpp (hunk 1 above)
Comment (mvieth) [reply test/octree/test_octree.cpp:1158]: Generally yes, but there are already several issues (and a whole "project") about better random generators in PCL.

**Review by kunaltyagi — APPROVED** [review]

**Pull request merged**